lib/Conversion/TorchOnnxToTorch/DefaultDomainGtoP.cpp

```diff
@@ -558,6 +558,95 @@ void mlir::torch::onnx_c::populateDefaultDomainGtoP(
                       binder.op, resultType, lhs, rhs);
                   return success();
                 });
+  patterns.onOp(
+      "GatherND", 13, [](OpBinder binder, ConversionPatternRewriter &rewriter) {
+        Torch::ValueTensorType resultType;
+        Value data, indices;
+        int64_t batch_dims;
+        if (binder.tensorOperandAtIndex(data, 0) ||
+            binder.tensorOperandAtIndex(indices, 1) ||
+            binder.tensorResultType(resultType) ||
+            binder.s64IntegerAttr(batch_dims, "batch_dims", 0))
+          return failure();
+        // TODO: Add support for non zero batch_dims.
+        if (batch_dims != 0)
+          return rewriter.notifyMatchFailure(
+              binder.op, "non zero batch_dims is not supported yet");
+        auto ctx = binder.op->getContext();
+        Location loc = binder.getLoc();
+        auto indicesTy = cast<Torch::ValueTensorType>(indices.getType());
+        auto dataTy = cast<Torch::ValueTensorType>(data.getType());
+        // TODO: Add support for dynamic sizes.
+        if (!dataTy || !dataTy.areAllSizesKnown())
+          return failure();
+        if (!indicesTy || !indicesTy.hasSizes())
+          return failure();
+
+        ArrayRef<int64_t> dataShape = dataTy.getSizes();
+        ArrayRef<int64_t> indicesShape = indicesTy.getSizes();
+        int64_t indicesRank = indicesShape.size();
+        int64_t indicesLastDim = indicesShape.back();
+        // Given data tensor of rank r >= 1, indices tensor of rank q >= 1, and
+        // batch_dims integer b, onnx.gather_nd gathers slices of data into an
+        // output tensor of rank q + r - indices_shape[-1] - 1 - b.
+        // indices_shape[-1] must be static to have deterministic output rank.
+        if (indicesLastDim == Torch::kUnknownSize)
+          return failure();
+
+        // Reduce indices to properly map it to the flattened data.
+        SmallVector<int64_t> indexMultiplier;
+        int64_t prod = 1;
+        for (int64_t i = indicesLastDim - 1; i >= 0; i--) {
+          indexMultiplier.push_back(prod);
+          prod *= dataShape[i];
+        }
+        std::reverse(indexMultiplier.begin(), indexMultiplier.end());
+        Value indexMultiplierList =
+            createConstantIntList(binder, rewriter, indexMultiplier);
+        Value none = rewriter.create<Torch::ConstantNoneOp>(loc);
+        Value boolFalse = rewriter.create<Torch::ConstantBoolOp>(loc, false);
+        Type multiplierTy =
+            Torch::ValueTensorType::get(ctx, llvm::ArrayRef({indicesLastDim}),
+                                        indicesTy.getOptionalDtype());
+        Value multiplierTensor = rewriter.create<Torch::AtenTensorOp>(
+            loc, multiplierTy, indexMultiplierList,
+            Torch::getDtypeIntValueForType(rewriter, loc,
+                                           indicesTy.getOptionalDtype()),
+            none, boolFalse);
+        Value updatedIndices = rewriter.create<Torch::AtenMulTensorOp>(
```

Comment:
When taking indice values, multiplying, and summing we are better to slice each out explicitly, mul, and ad. We want to avoid using reduction style cases as they form more complex dispatches that often do not work with fusion.